### PR TITLE
Add secret management for dotnet-status-web

### PR DESCRIPTION
**Pull request by alexperovich**

### CODE REVIEW

**Comment by ChadNedzlek** [comment]
Having to duplicate all the data between staging and prod is... really unfortunate.  I guess I never realized that would be a thing when looking at the designs of the stuff.  Is there some way that can be avoided?

**Comment by alexperovich** [comment]
I could add some kind of "import" type thing in the yaml if that makes sense. It means the secret names would have to be a bit more generic. The staging and prod vaults with both have "the-storage-account" as a secret and it would point to different storage accounts.

Talking implementation, it could be a raw text #include style, or an explicit property that says "there are some secrets in this file, add them too." I think I prefer the latter.

File: src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs [hunk 1]
```diff
@@ -93,9 +100,10 @@ public override async Task<List<SecretData>> RotateValues(Parameters parameters,
             DateTimeOffset rollOn = _clock.UtcNow.AddMonths(6);
 
             return new List<SecretData> {
-                    new SecretData(appId, DateTimeOffset.MaxValue, DateTimeOffset.MaxValue),
+                    new SecretData(appId, DateTimeOffset.MaxValue, rollOn),
```
Comment (jakubstilec) [inline src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs:103]: Shouldn't it be max value as it never changes?

File: src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs [hunk 2]
```diff
@@ -14,19 +14,19 @@ public class GitHubAppSecret : SecretType<GitHubAppSecret.Parameters>
...
         private const string AppId = "-app-id";
         private const string AppPrivateKey = "-app-private-key";
-        private const string AppSecret = "-app-secret";
-        private const string AppHook = "-app-hook";
+        private const string OAuthId = "-oauth-id";
+        private const string OAuthSecret = "-oauth-secret";
+        private const string AppHookSecret = "-app-webhook-secret";
 
-        private readonly List<string> _suffixes = new List<string> { AppId, AppPrivateKey, AppSecret, AppHook };
+        private readonly List<string> _suffixes = new List<string> { AppId, AppPrivateKey, OAuthId, OAuthSecret, AppHookSecret };
```
Comment (jakubstilec) [inline src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs:29]: It looks like in status web it's somehow mixed or named differently - I created  GitHubAppSecret and GitHubOAuthSecret to differentiate between app and oauth. But it's probably easier to have it in one secret type. I think we can now remove GitHubOAuthSecret.

File: .vault-config/dotneteng-status-local.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    type: github-app-secret
+    parameters:
+      hasPrivateKey: true
+      hasWebhookSecret: true
+      hasOAuthSecret: true
+
+  grafana-api-token:
+    type: grafana-api-key
+    parameters:
+      environment: dotnet-eng-grafana-staging.westus2.cloudapp.azure.com
```
[the remark this answers is not in the document]
Comment (alexperovich) [reply .vault-config/dotneteng-status-local.yaml:27]: The more I look at it, the less I like the "production" and "staging" values here. If I am a random dev looking at this file, and I see "production," I have no way of knowing what that is. I would have to look at the source code for the secret manager to know what "production" means, and I would have to know to look there.

It also means that a new instance of grafana or helix or something else requires code changes to the secret manager, and I don't like that.

File: src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs (hunk 1 above)
Comment (alexperovich) [reply src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs:103]: That is technically correct, but but giving these values different dates doesn't really do anything, they all get rotated when the earliest one says to.

File: src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs (hunk 2 above)
Comment (alexperovich) [reply src/Microsoft.DncEng.SecretManager/SecretTypes/GitHubAppSecret.cs:29]: Ahh yea, I didn't see that other one. That makes sense. I don't know if we want to support old-style oauth apps, but github apps contain both oauth stuff, and github app stuff so they need both.

File: .vault-config/dotneteng-status-secrets.yaml [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  parameters:
+    connectionString: storage-account
+    permissions: racwdx
+    container: site
+    blob: keys.xml
+
+fr-teams-channel-webhook-url:
+  type: text
+  parameters:
+    description: The teams "incomming webhook" connector url for the FR issue notification channel
```
Comment (MattGal) [inline .vault-config/dotneteng-status-secrets.yaml:33]: nit:  Incoming

File: src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+using Microsoft.DncEng.CommandLineLib;
+
+namespace Microsoft.DncEng.SecretManager.SecretTypes
+{
+    [Name("zenhub-access-token")]
+    public class ZenHubAccessToken : GenericAccessToken
+    {
+        private readonly string[] _expirationDateFormats = new[] { "yyyy-MM-dd", "yyyy-MM-dd HH:mm:ss" };
+
+        protected override string HelpMessage => "Please login to https://app.zenhub.com/dashboard/tokens using your GitHub account and create a new token.";
```
Comment (MattGal) [inline src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs:14]: Does the user need group membership or other stuff for their GH account to do this?  May be worth mentioning.

File: src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
+using System;
```
Comment (MattGal) [inline src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs:1]: header needed

File: src/Microsoft.DncEng.SecretManager/SecretTypes/KustoConnectionString.cs [hunk 7]
```diff
@@ -1,6 +1,6 @@
+using System.Text;
```
Comment (MattGal) [inline src/Microsoft.DncEng.SecretManager/SecretTypes/KustoConnectionString.cs:1]: header needed

**Review by MattGal — APPROVED** [review]
LGTM, I like the self-explanatory-ness of the yaml format.

File: src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs (hunk 5 above)
Comment (alexperovich) [reply src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs:14]: I do not know. They do need to be able to use zenhub for what we need. So I assume just "be in our team"

File: src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs (hunk 6 above)
Comment (alexperovich) [reply src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs:1]: most files in this repo don't have the headers. I'd rather add them all in one separate PR, than just add a few here.

Comment (ChadNedzlek) [reply src/Microsoft.DncEng.SecretManager/SecretTypes/ZenHubAccessToken.cs:1]: Yeah, I think a lot of this was me.  I should be able to have ReSharper do it all in one fell swoop, I think...  should I make a PR for that? Should be a fun merge conflict.

**Review by jakubstilec — APPROVED** [review]

**Pull request merged**